chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,177 @@
+
+The Journal nodes allow for reconstruction of the file system even in the case of disk falure of the Active name node and the Standby namenode by saving the Edit logs which can be used to reconstruct the FSImage in the case of hardware failure or losing even both the namenodes, using a cluster of journal nodes which all of the logs are written to all of them to allow for FSImage reconstruction even in the case a journal node fails.
+(it is possible to have both a standby node and a secondary node but this is not needed because the standby node preforms checkpoints of the FSImage in intervale, so it would be a waste of resources)
+
+So when a namenode crashes Zookeeper failover controller detects it and marks the Node as unhealthy, when the namenode crashed it also releases the ephermal node due to the session closing becaused no ping was sent.
+Then the other namenodes from the zookeeper quorum then try to create an ephermal node, and the first one to do so aquires the lock, and becomes the Active namenode, only one Node can own the lock at a time so it guarantees that only one active namenode exists at a time.
+
+in the case of a split brain, the journal nodes will allow only one name node to have the role to write them at a time which removes the possible risk of data loss during a split brain.
+
+because the jouurnal nodes allow only one name node to have the role of writing to them at a time, in the case of the failover, the newly elected active name node only can write to the journal nodes which means that the other namenode won't write anything or in other words, won't be active as it dissallows it to be active, which in turn, allows for the new namenode to continue operations safely.
+
+and in the case of failover, the new Name node will read from the journal nodes and sync the FSImage so that no data would be lost.
+
+If HA is not turned on (default) there is no standby namenode and instead there is a secondary namenode which all it does is read the edit logs and build the FSImage so that if the Namenode falls, it won't have to rebuild the FSimage, which can be a heavy task and cause it to fail again or increase the downtime.
+
+In the case of name node failure, the cluster becomes unavailable up until the namenode restarts.
+
+
 7. **Q:** What is the role of the NameNode in HDFS?
+7. **A:** The role of the NameNode is to save metadata about the file and to manage where they are saved, in addition to managing and saving the FS structure and also replicate / delete replications as needed.
+the metadata is stored in memory and also in a persistant FSImage file, in addition to a local editlog which includes the logs of all the eddits that have happened to the file system structure.
+the namenodes also monitors all the data nodes and knows which are alive and which are down.
+
 8. **Q:** What is a DataNode in HDFS?
```

Comment:
So this question is actually different in the original hafifot repo. Please answer it as well (Q. 23):
https://github.com/883G/Onboarding-Newbies/blob/main/chapter_01/day_04.md#chapter-5-hadoop-distributed-file-system-hdfs